### PR TITLE
Refactor connector opsdroid pointer

### LINKED ISSUE #725: Refactor connector opsdroid pointer
Currently in the `Connector` class we expect opsdroid to pass a pointer to itself on `connect` and `disconnect`. This seems a little pointless as the connector should be storing the pointer as an attribute anyway.

### Steps

- Add `opsdroid` as a kwarg in the `__init__` method of the `Connector` class and store it as an attribute. 
- Remove the opsdroid kwarg from the `connect` and `disconnect` method as they are no longer needed.
- All core connectors would also need updating to reflect this.

### PR DESCRIPTION
**Pull request by zonagilreath**
# Description

Slight update to previous request

Fixes #725 

## Status
**READY** 

## Type of change

_Please delete options that are not relevant._

- Bug fix (non-breaking change which fixes an issue)
- Documentation (fix or adds documentation)

# How Has This Been Tested?

Tox tests passing

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by FabioRosado** [comment]
I've created a PR in your fork with the changes that I tried to push 👍

**Comment by zonagilreath** [comment]
@FabioRosado I'm sorry I didn't see that before, last week was pretty busy with work and the holiday. Do I still need to merge that PR, or is there anything else I need to do now?

**Comment by FabioRosado** [comment]
Hello @zonagilreath  my apologies for not replying yesterday. Yeah that PR still needs to be merged into your opsdroid fork - for what tox told me everything should pass once that PR is merged and I'm sure this one will be updated as soon as you merge my PR into yours then we can see if we can get this merged without any issues 😄 👍

**Comment by zonagilreath** [comment]
Okay, done. But there's a conflict on the listen method of each of the connectors, and in a few of the tests and I'm not sure how they should be resolved.

**Comment by FabioRosado** [comment]
Yeah this week we changed a few things related to the logging and connectors, I have fixed the conflicts for you 😄 👍

**Comment by zonagilreath** [comment]
Great, hope everything is resolved now

**Comment by FabioRosado** [comment]
Hmmm this is odd everything is broken now but my tox was showing green o.O

**Comment by zonagilreath** [comment]
Yeah, it looks like one issue is that the connector.listen() is still getting a None argument, which I thought I had fixed. Should I get that fixed?

**Comment by FabioRosado** [comment]
I didn't have time to check exactly what was wrong here but if you have the time that would be great 👍

**Comment by zonagilreath** [comment]
I'm about to test the rest, but there's one error I don't understand:
`___________________ TestConnectorSlackAsync.test_listen_loop ___________________
self = <test_connector_slack.TestConnectorSlackAsync testMethod=test_listen_loop>
    async def test_listen_loop(self):
        """Test that listening consumes from the socket."""
        connector = ConnectorSlack({"api-token": "abc123"}, opsdroid=OpsDroid())
        connector.receive_from_websocket = amock.CoroutineMock()
        connector.receive_from_websocket.side_effect = Exception()
        with self.assertRaises(Exception):
            await connector.listen(amock.CoroutineMock())
>       self.assertTrue(connector.receive_from_websocket.called)
E       AssertionError: False is not true
tests/test_connector_slack.py:102: AssertionError`

**Comment by FabioRosado** [comment]
That `amock.CoroutineMock()` shouldnt be there. Im assuming that if we take that bit out it will work.

Usually we either call opsdroid with `with OpsDroid` or use `opsdroid = amock.CoroutineMock()` so that test sort of uses a short hand for opsdroid

**Comment by zonagilreath** [comment]
Oh okay. I was wrong earlier about having fixed the arguments for listen() in all the tests. I hadn't changed the signature of listen() in the first place, I think that came in when I merged your PR.

**Comment by zonagilreath** [comment]
Now everything is passing but this: 
```
____________________ TestConnectorGitHubAsync.test_connect ____________________

self = <test_connector_github.TestConnectorGitHubAsync testMethod=test_connect>

    async def test_connect(self):
        opsdroid_mock = amock.CoroutineMock()
        with amock.patch('aiohttp.ClientSession.get') as patched_request:
            mockresponse = amock.CoroutineMock()
            mockresponse.status = 200
            mockresponse.json = amock.CoroutineMock(return_value={
                "login": 'opsdroid'
            })
            patched_request.return_value = asyncio.Future()
            patched_request.return_value.set_result(mockresponse)
            await self.connector.connect()
            self.assertEqual(self.connector.github_username, "opsdroid")
>           self.assertTrue(opsdroid_mock.web_server.web_app.router.add_post.called)
E           AssertionError: False is not true

tests\test_connector_github.py:61: AssertionError
```

**Comment by FabioRosado** [comment]
Yeah I changed the listen method since opsdroid was being passed as an argument for the class, I thought I had changed all the listen texts since tox didn't complain on my end but I'm not sure now

**Comment by zonagilreath** [comment]
Yeah that's weird that they weren't caught by the tests before. Got them fixed now, but I don't understand what's supposed to be happening with the test well enough to debug. I'll go ahead and push what I have, but if you see what needs to be fixed I'm happy to do it.

**Comment by FabioRosado** [comment]
Just to let you know that I'm off today and I plan to spend some time with this PR to see why that test is now failing, will let you know if I can fix or offer any sort of solution for it 👍

**Comment by jacobtomlinson** [comment]
I would check the Facebook connector tests to see what we did there, as they are both adding routes to the web server. Otherwise I'll try and take a look soon.

**Comment by zonagilreath** [comment]
I got it, I think.

**Comment by zonagilreath** [comment]
Still failing on docker commands, but everything else is working now.

File: .vs/VSWorkspaceState.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (FabioRosado) [inline .vs/VSWorkspaceState.json]: Do we need this file?

Comment (zonagilreath) [reply .vs/VSWorkspaceState.json]: Nope, sorry, wasn't thinking and ran the wrong add command

File: opsdroid/connector/websocket/__init__.py [hunk 2]
```diff
@@ -18,13 +18,12 @@
...
 
-    def __init__(self, config):
+    def __init__(self, config, opsdroid=None):
         """Create the connector."""
-        super().__init__(config)
+        super().__init__(config, opsdroid=opsdroid)
         _LOGGER.debug("Starting Websocket connector")
         self.name = "websocket"
         self.config = config
-        self.opsdroid = None
```
Comment (FabioRosado) [inline opsdroid/connector/websocket/__init__.py]: Do you know this is the reason why we are getting the `NoneType` issue within docker?

File: .vs/VSWorkspaceState.json (hunk 1 above)
Comment (zonagilreath) [reply .vs/VSWorkspaceState.json]: Should be good now

File: opsdroid/connector/websocket/__init__.py (hunk 2 above)
Comment (zonagilreath) [reply opsdroid/connector/websocket/__init__.py]: I bet that wherever `ConnectorWebsocket` is being initialized it's not being given the opsdroid argument that it needs now, so it's defaulting to None.

Comment (zonagilreath) [reply opsdroid/connector/websocket/__init__.py]: I think I found it in core

**Comment by zonagilreath** [comment]
Ah, I thought it was `OpsDroid.start_connectors()` that wasn't liking the new argument but I see now. Updating the mock connectors, and will test again.

**Comment by zonagilreath** [comment]
Woo!

**Comment by FabioRosado** [comment]
This is an amazing work! Congratulations 🎉 
Also well done finding those buggers that were making the tests complaining 😄 👍 

I am going to merge this now!

We also give stickers to contributors so if you would like them DM your home address to [opsdroid twitter account](https://twitter.com/opsdroid) and Jacob will send them to you 👍

**Pull request merged**